### PR TITLE
Fix: Enforce application deadline on topics

### PR DESCRIPTION
**Pull request by bensofficial**
## Summary
- Add server-side validation in `ApplicationService.createApplication()` to reject applications after the topic's deadline has passed
- Hide or replace "Apply" buttons on the frontend (TopicPage, TopicCard, CollapsibleTopicElement) when the deadline has passed
- Add `applicationDeadline` to `TopicOverviewDto` so the landing page can check deadlines without fetching full topic details

## Test plan
- [ ] Create a topic with an application deadline in the past and verify that the "Apply" button is not shown
- [ ] Verify that submitting an application via API to a topic with a past deadline returns an error
- [ ] Create a topic with a future deadline and verify that applications can still be submitted
- [ ] Create a topic without a deadline and verify that applications work as before

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Topics can now have application deadlines; applications are blocked after a deadline passes.
  * UI updates: expired topics show "More Information" instead of "Apply" across listing views and the topic page, and the "Apply Now" action is disabled when a deadline has passed.

* **Tests**
  * Added integration and cron tests covering deadline enforcement and boundary conditions.

### CODE REVIEW

**Review by Claudia-Anthropica — APPROVED** [review]
@bensofficial Clean and well-structured fix. Server-side validation follows the existing `closedAt` pattern nicely, the DTO update is minimal, and the frontend changes are consistent across all three components. Nice work.

File: server/src/test/java/de/tum/cit/aet/thesis/service/ApplicationServiceIntegrationTest.java [hunk 1]
```diff
@@ -309,6 +309,121 @@ void rejectList_RecentApplications_NotRejected() throws Exception {
...
+						.executeUpdate();
+				entityManager.clear();
+			});
+
+			String studentAuth = createRandomAuthentication("student");
+			mockMvc.perform(MockMvcRequestBuilders.post("/v2/applications")
+							.header("Authorization", studentAuth)
+							.contentType(MediaType.APPLICATION_JSON)
+							.content(objectMapper.writeValueAsString(applyTo(topicId))))
+					.andExpect(status().isBadRequest());
```
Comment (Claudia-Anthropica) [inline server/src/test/java/de/tum/cit/aet/thesis/service/ApplicationServiceIntegrationTest.java:454]: @bensofficial [nit] The name `DeadlineExactlyNow_Rejected` (and the commit message wording) doesn't match what's actually tested — you set the deadline 1 ms in the past, so this is effectively a duplicate of the `PastDeadline` case. Real "exactly now" with `Instant.now(clock).isAfter(deadline)` would actually be accepted (`isAfter` is strict). Either rename to `JustPastDeadline_Rejected`, or use the `MutableClock` (per the commit message's stated intent) to pin time and assert that `now == deadline` is in fact accepted while `deadline - 1ns` is rejected.

<details><summary>🤖 Prompt for AI agents</summary>

In `server/src/test/java/de/tum/cit/aet/thesis/service/ApplicationServiceIntegrationTest.java`, the test `createApplication_DeadlineExactlyNow_Rejected` (around line 402) does not actually exercise the `now == deadline` boundary — it sets the deadline 1 ms in the past, which duplicates the `PastDeadline` case. Either rename it to `createApplication_JustPastDeadline_Rejected` and update the comment, or rewrite it to inject the `MutableClock` test bean, pin it to a fixed instant `T`, set the topic's deadline to exactly `T`, and assert the request succeeds (since `isAfter` is strict), then add a second case where the deadline is `T - 1 nanosecond` and assert it's rejected.

</details>

**Review by Claudia-Anthropica — APPROVED** [review]
@bensofficial Nice follow-up — the Clock injection is clean, the new integration tests cover past/future/null deadline well, and the cron test fix has a clear explanation for the two-step setup. One small naming nit inline, otherwise good to go.

File: server/src/test/java/de/tum/cit/aet/thesis/service/ApplicationServiceIntegrationTest.java (hunk 1 above)
Comment (krusche) [reply server/src/test/java/de/tum/cit/aet/thesis/service/ApplicationServiceIntegrationTest.java:454]: Fixed in 917ccbc0: rewrote the test to actually exercise the boundary. Pinned `MutableClock` to T, set the deadline to T, and asserted rejection (paired with a new `createApplication_DeadlineJustAfterNow_Succeeds` for the T+1ms case). The service-side check was also tightened from `isAfter` to `!isBefore` so the equality boundary genuinely belongs in the rejected case.

File: client/src/pages/ReplaceApplicationPage/components/SelectTopicStep/components/CollapsibleTopicElement.tsx [hunk 2]
```diff
@@ -146,15 +146,21 @@ const CollapsibleTopicElement = ({ topic, onApply }: ICollapsibleTopicElementPro
...
+                new Date(fullTopic.applicationDeadline) < new Date()
+              ) && (
+                <Button
+                  onClick={() => onApply(fullTopic || undefined)}
+                  fullWidth
+                  disabled={!fullTopic}
+                >
+                  Apply
+                </Button>
+              )}
```
[the remark this answers is not in the document]
Comment (krusche) [reply client/src/pages/ReplaceApplicationPage/components/SelectTopicStep/components/CollapsibleTopicElement.tsx]: Fixed in 917ccbc0: when the deadline has passed, both `CollapsibleTopicElement` and `TopicPage` now show a disabled Apply button with an "Application deadline has passed." caption instead of hiding it.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ApplicationService.java [hunk 3]
```diff
@@ -200,6 +206,10 @@ public Application createApplication(User user, UUID researchGroupId, UUID topic
 			throw new ResourceInvalidParametersException("This topic is already closed. You cannot submit new applications for it.");
 		}
 
+		if (topic != null && topic.getApplicationDeadline() != null && Instant.now(clock).isAfter(topic.getApplicationDeadline())) {
+			throw new ResourceInvalidParametersException("The application deadline for this topic has passed. You cannot submit new applications for it.");
+		}
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ApplicationService.java]: Fixed in 917ccbc0: switched to `!now.isBefore(deadline)` so `now == deadline` is rejected, and consolidated on a single `Instant.now(clock)` for the deadline check, `createdAt`, and `consentTimestamp` instead of mixing in `Instant.now()`.

**Review by krusche — APPROVED** [review]

**Review by Claudia-Anthropica — APPROVED** [review]
@bensofficial All earlier feedback addressed cleanly — the boundary test now genuinely exercises `now == deadline` with `MutableClock`, the disabled-button UX is consistent across topic views, and the unrelated maintenance fixes (AvatarController 404, build.gradle worktree support, e2e reliability) are well-scoped. Approving.
